src/TPLinkManager.php

```diff
@@ -20,11 +21,47 @@ public function device($name = 'default')
             throw new InvalidArgumentException('You have not setup the details for a device named ' . $name);
         }
 
-        return $this->newTPLinkDevice($this->config[$name], $name);
+        return new TPLinkDevice($this->config[$name], $name);
     }
 
     protected function newTPLinkDevice($config, $name)
     {
+        $this->config[$name] = $config;
         return new TPLinkDevice($config, $name);
     }
-}
\ No newline at end of file
+
+    public function deviceList()
+    {
+        return $this->config;
+    }
+
+    public function autoDiscoverTPLinkDevices($ipRange, $timeout = 1)
+    {
```

Comment:
I'm a big fan of Adam Wathan's idea of [Refactoring to Collections](https://adamwathan.me/refactoring-to-collections/).

I try to not use foreach loops in the code (I notice there is one actually I forgot to refactor, so I'll do that soon). I'd like to refactor this function and separate it out a little more.

If you're in the mood to have a look at this again that would be great otherwise I'll have a bash at it later.